psfws/gen_params.py

```diff
@@ -0,0 +1,371 @@
+        The u/v components of velocity correspond to north/south winds,
+        respectively, and the wind direction is given as degrees west of north.
+
+        """
+        speed = np.hstack([self.telemetry['speed'][pt],
+                           self.gfs_winds.iloc[pt]['speed'][self._gfs_stop:]])
+
+        direction = np.hstack([self.telemetry['dir'][pt],
+                               self.gfs_winds.iloc[pt]['dir'][self._gfs_stop:]]
+                              )
+
+        u = np.hstack([self.telemetry['u'][pt],
+                       self.gfs_winds.iloc[pt]['u'][self._gfs_stop:]])
+
+        v = np.hstack([self.telemetry['v'][pt],
+                       self.gfs_winds.iloc[pt]['v'][self._gfs_stop:]])
+
+        height = np.hstack([self.h0, self.h_gfs[self._gfs_stop:]])
+
+        return {'u': u, 'v': v, 'speed': speed,
+                'direction': utils.smooth_direction(direction), 'h': height}
+
+    def _interpolate_wind(self, p_dict, h_out, kind='gp'):
+        """Return new wind dict, vals interpolated to new heights h_out."""
+        new_u = utils.interpolate(p_dict['h'], p_dict['u'], h_out, kind)
+        new_v = utils.interpolate(p_dict['h'], p_dict['v'], h_out, kind)
+
+        new_direction = utils.to_direction(new_v, new_u)
+
+        return {'u': new_u, 'v': new_v, 'speed': np.hypot(new_u, new_v),
+                'direction': utils.smooth_direction(new_direction), 'h': h_out}
+
+    def _draw_ground_cn2(self):
+        """Return ground Cn2 model, with parameters randomly drawn.
+
+        Model from "where is surface layer turbulence" (Tokovinin 2010), params
+        drawn from normal distribution around values from paper.
+        """
+        # model valid till 300m
+        h_gl = np.linspace(self.h_dome, 300, 100)
+
+        tokovinin_a = [-13.38, -13.15, -13.51]
+        tokovinin_p = [-1.16, -1.17, -0.97]
+
+        # adjust by sqrt(N-1) for sample variance
+        a = self._rng.normal(loc=np.mean(tokovinin_a),
+                             scale=np.std(tokovinin_a)/np.sqrt(2))
+        p = self._rng.normal(loc=np.mean(tokovinin_p),
+                             scale=np.std(tokovinin_p)/np.sqrt(2))
+
+        # also return h of GL model in km, adjusted to observatory height
+        return utils.ground_cn2_model({'A': a, 'p': p},
+                                      h_gl), self.h0 + h_gl/1000
+
+    def get_cn2(self, pt):
+        """Get Cn2 and h arrays for datapoint with index pt.
+
+        Use Hufnagel model and GFS winds to calculate a Cn2 profile above 3km,
+        stacked with a ground layer model draw.
+        """
+        # pick out relevant wind data
+        raw_winds = self.get_raw_wind(pt)
+
+        # make a vector of heights where hufnagel will be valid
+        h_huf = np.linspace(self.h0 + 3, max(self.h_gfs), 100)
+        # interpolate wind data to those heights
+        speed_huf = self._interpolate_wind(raw_winds, h_huf)['speed'].flatten()
```

Comment:
Thanks for these thoughts. I've ended up removing the option to change interpolation methods since it doesn't have a meaningful impact; all is done with scipy cubic. (comment resolves a few conversations below, as well)